File: DailyCoding/string_permutation/string_perm.py
from typing import List

def find_combs(s: str) ->List[str]:
    if len(s) == 0:
        return [""]
    else:
        combs_without_first = find_combs(s[1:])
        combs_with_first = [s[0]+comb for comb in combs_without_first]
        return combs_with_first + combs_without_first
    
s = 'abc'
print(f"perms: {find_combs(s)}")

File: DailyCoding/string_permutation/string_perm.cpp
#include <vector>
#include <string>
#include <iostream>



void find_combs_util(const std::string& s, std::vector<std::string>& result)
{
    if(s.empty())
    {
        result.push_back("");
        return;
    }
    else
    {
        std::vector<std::string> combs_without_first;
        find_combs_util(s.substr(1),combs_without_first);

        for(auto const & t: combs_without_first)
        {
            result.push_back(t);
            result.push_back(s[0]+t);
        }
        return;
    }
}

std::vector<std::string> finds_comb(const std::string& s)
{
    std::vector<std::string> result;
    find_combs_util(s, result);
    return result;
}

int main()
{
    std::string s = "abc";
    std::vector<std::string> r = finds_comb(s);
    for(auto const &comb : r)
    {
        std::cout << "\'" << comb << "\'" << ",";
    }
    std::cout << "\n";
    std::cout << "len: " << r.size() << "\n";
    return 0;
}